test/runtime/runtime_plugins_test.go

```diff
@@ -0,0 +1,370 @@
+// Copyright 2018 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// +build linux,cgo darwin,cgo
+
+// This exists in a separate testing package for the following reason: https://github.com/golang/go/issues/17928
+
+package runtime
+
+import (
+	"context"
+	"fmt"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/runtime"
+	"github.com/open-policy-agent/opa/types"
+	"github.com/open-policy-agent/opa/util/test"
+	"net/http"
+	"net/http/httptest"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func makeBuiltinWithName(name string) string {
+	return fmt.Sprintf(`
+package main
+
+import (
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/types"
+	"github.com/open-policy-agent/opa/topdown"
+)
+
+var TruthfulBuiltin = &ast.Builtin{
+	Name: "%v",
+	Decl: types.NewFunction(
+		types.Args(types.N, types.N),
+		types.B,
+	),
+}
+
+func Truthful(a, b ast.Value) (ast.Value, error) {
+	return ast.Boolean(true), nil
+}
+
+func Init() error {
+	ast.RegisterBuiltin(TruthfulBuiltin)
+	topdown.RegisterFunctionalBuiltin2(TruthfulBuiltin.Name, Truthful)
+	return nil
+}
+`, name)
+}
+
+// returns go code for a plugin named name that makes a single get request to url upon start
+func getPluginWithNameAndURL(name, url string) string {
+	return fmt.Sprintf(
+		`
+package main
+
+import (
+	"context"
+	"github.com/open-policy-agent/opa/plugins"
+	"github.com/open-policy-agent/opa/runtime"
+	"net/http"
+)
+
+type Tester struct{}
+
+func (t *Tester) Start(ctx context.Context) error {
+	_, err := http.Get("%v")
+	return err
+}
+
+func (t *Tester) Stop(ctx context.Context) {
+	return
+}
+
+func InitTester(m *plugins.Manager, config []byte) (plugins.Plugin, error) {
+	return &Tester{}, nil
+}
+
+func Init() error {
+	runtime.RegisterPlugin("%v", InitTester)
+	return nil
+}
+`, url, name)
+}
+
+func TestRegisterBuiltinSingle(t *testing.T) {
```

Comment:
We don't have to gate the PR on this but I noticed it takes ~1 second to shell out and build each plugin. This adds up to around ~12 seconds to test the entire package. It seems like (a) testing the recursive listing of shared objects separately from the building and (b) updating the test to use TestMain and build all of the required shared objects once up front would improve this quite a bit.